### ISSUE TITLE
didn't get call transfer audio to work in Android

### ISSUE THREAD

**Issue by juha-h** [alfredh/baresip#548]
Call transfer based on UA_EVENT_CALL_TRANSFER works fine in Linux baresip.  SIP signaling of call transfer works OK also in Android, but after the transfer is complete, Android UA does not send any RTP audio to transfer destination.  Perhaps it cannot record audio to two clients (original caller that made the transfer request and transfer destination) at the same time.  This would be needed, since the original call leg is not closed yet when the new leg to transfer destination is established.

In Android log I see the following at the time when the new call leg to transfer destination is established:
```
09-30 09:51:55.429 20556 20660 D Baresip : ua event (CALL_RINGING) sip:jh@test.tutpro.com
09-30 09:51:55.453 20556 20660 I libOpenSLES: Emulating old channel mask behavior (ignoring positional mask 0x4, using default mask 0x1 based on channel count of 1)
09-30 09:51:55.455 20556 20660 D AudioTrack: Client defaulted notificationFrames to 109 for frameCount 328
09-30 09:51:55.456  3553 26543 D APM_AudioPolicyManager: startOutput() output 13, stream 3, session 2793
09-30 09:51:56.796 20556 20660 D Baresip : ua event (CALL_RTCP) audio
09-30 09:51:57.854 20556 20661 D Baresip : audio: Set audio decoder: opus 48000Hz 2ch
09-30 09:51:57.856 20556 20660 I libOpenSLES: Emulating old channel mask behavior (ignoring positional mask 0x3, using default mask 0x3 based on channel count of 2)
09-30 09:51:57.857 20556 20660 D AudioTrack: Client defaulted notificationFrames to 962 for frameCount 1924
09-30 09:51:57.859  3553  3856 D APM_AudioPolicyManager: startOutput() output 13, stream 3, session 2801
09-30 09:51:57.860 20556 20661 D Baresip : audio: player started with sample format S16LE
09-30 09:51:57.860 20556 20661 D Baresip : audio: Set audio encoder: opus 48000Hz 2ch
09-30 09:51:57.862 20556 20660 I libOpenSLES: Emulating old channel mask behavior (ignoring positional mask 0x3, using default mask 0xc based on channel count of 2)
09-30 09:51:57.863  3553 25675 D audio_hw_primary: adev_open_input_stream: enter: sample_rate(48000) channel_mask(0xc) devices(0x80000004)        stream_handle(0xa6435b00) io_handle(686)
09-30 09:51:57.866  3553 23228 I AudioFlinger: AudioFlinger's thread 0xa3d83180 ready to run
09-30 09:51:57.866  3553 23228 D audio_hw_primary: in_standby: enter: stream (0xa6435b00) usecase(7: audio-record)
09-30 09:51:57.870  3553  3856 W AudioFlinger: createRecordTrack_l(): mismatch between requested flags (00000005) and input flags (00000001)
09-30 09:51:57.870  3553 23228 D audio_hw_primary: in_standby: enter: stream (0xa6435b00) usecase(7: audio-record)
09-30 09:51:57.873 20556 20660 I AudioRecord: AUDIO_INPUT_FLAG_FAST successful; frameCount 0
*** 09-30 09:51:57.874  3553 26543 E APM_AudioPolicyManager: startInput(686) failed: other input 678 already started
*** 09-30 09:51:57.874 20556 20660 E AudioRecord: start() status -38
09-30 09:51:57.875 20556 20661 D Baresip : audio: source started with sample format S16LE
09-30 09:51:57.875 20556 20661 D Baresip : audio tx pipeline:    opensles ---> opus
09-30 09:51:57.875 20556 20661 D Baresip : audio rx pipeline:    opensles <--- opus
09-30 09:51:57.876 20556 20661 D Baresip : foo@test.tutpro.com: Call established: sip:jh@test.tutpro.com
09-30 09:51:57.876 20556 20660 D Baresip : ua event (CALL_ESTABLISHED) sip:jh@test.tutpro.com
```
See the lines prefixed with ***.

I'm ready to give up on this, but just thought to let people know what happened.  I'll close this issue in a couple of days.

**Comment by juha-h** [comment]
Looks like there is google issue on this: https://issuetracker.google.com/issues/36988922 (requires google login).  One user comments:

> Following previous comment, after investigation (looking the source code of AudioPolicyManagermainly, but also developers issues and some tests), this is clearly because _only one_ app can record at same time. Multiple or sharing input sources are not supported.

I have not tried yet if it would be somehow possible free recording from the first call leg, before calling call_connect() on the second leg.  Would audio_mute() or audio_set_hold() release audio input source?

**Comment by juha-h** [comment]
Calling audio_mute() or audio_set_hold() didn't help, but audio_stop() did it!  Currently I thus have:
```
    LOGD("transfering call %s to %s\n", native_call, native_uri);
    re_thread_enter();
    err = ua_call_alloc(&call2, ua, VIDMODE_OFF, NULL, call, call_localuri(call), true);
    if (!err) {
        struct pl pl;
        pl_set_str(&pl, native_uri);
        audio_stop(call_audio(call));
        err = call_connect(call2, &pl);
        ...
```
But what if the second leg call to the transfer target fails, is there a way to somehow start the stopped first leg audio again?

**Comment by juha-h** [comment]
Since audio_set_hold() does not release source audio device, it seems likely that audio would not work in the new call that is made after putting the first call on hold.

Perhaps it would be possible to automatically release audio/video source when call is put on hold and then automatically acquire it again when call is released from hold?

**Comment by alfredh** [comment]
hi @juha-h 

I think this is a very special usecase ( Android *and* call transfer) so I would not
waste too much time on it.

but in theory it should be possible to start and stop the audio
while you are doing the call transfer process:


```
int  audio_start(struct audio *a)
void audio_stop(struct audio *a)
```

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> I think this is a very special usecase ( Android *and* call transfer)
> so I would not waste too much time on it.

I just though to give it a try.

> but in theory it should be possible to start and stop the audio
> while you are doing the call transfer process:
> 
> 
> ```
> int  audio_start(struct audio *a)
> void audio_stop(struct audio *a)
> ```

Yes, except that audio_start() is not in the API.

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> if it is needed, we can move `audio_start` to the public API :)

Please do that and I'll give audio_stop/audio_start a try and then close
this issue.

**Comment by alfredh** [comment]
okay, fixed in commit 61e7bd4d9b7a32d88877a624f268a57f5fba113a
